### ISSUE TITLE
Transformer: need flags for head_count and hidden_size for transformer models

### ISSUE THREAD

**Issue by David-Levinthal** [OpenNMT/OpenNMT-py#771]
The transformer code can only run the base Transformer model as head_count and (feed forward) hidden_size are only set as defaults (in multiple places) and are never actually passed in as arguments
Currently to change the model size one must edit the source code.
the initializations/defaults are currently set in two places
 in omnt/modules/Transformer.py at line 57
    def __init__(self, size, dropout,
                 head_count=8, hidden_size=2048):
and line 153
    def __init__(self, size, dropout,
                 head_count=8, hidden_size=2048):
and I believe the head_count is hardcoded in an assertion in onmt/modules/MultiHeadedAttn.py line 97
        aeq(self.model_dim % 8, 0)
The high accuracy translation actually requires using the "Big" model in Tensorflow t2t or the tutorial and now in MLPerf/Translation/tensoflow/transformer which has an embedding/model_size of 1024, 16 heads and a feed forward hidden size of 4096

It would be nice if Transformer.py and MulitHeadedTransformer did not use the variable name hidden_size to sometimes mean the model_size   :-)
see Class TransformerEncoder(EncoderBase)
    Args:
       num_layers (int): number of encoder layers
       hidden_size (int): number of hidden units
but it is invoked with the second argument being opt.rnn_size at line 64 of ModelConstructor.py
and maybe should actually be opt.word_vec_size?

**Comment by vince62s** [comment]
I agree, while writing #769 I went into the same thing.
If this is merged, I will also adjust and take this into account.
Never the less, several discussions in T2T gitter suggest that when training a very long time, the base model meets almost the big one in terms of performance.
Right now I am more concerned by decoding speed.

**Comment by lpentecost** [comment]
While this doesn't address the issue of variable name `hidden_size` inside model params vs. `TransformerEncoderLayer` definition or the repeated default value inits (so I'm not initiating a pull request), a quick work-around to parameterize the Transformer is to make the following 3 changes:<br/>
<br/>

1.  Add `ff_size` and `head_count` flags in model param CLAs for feed forward hidden size and number of heads, respectively, in onmt/opts.py line 72:  <br/>
<br/>

    `group.add_argument('-ff_size', type=int, default=2048,
                       help='Hidden size for feed forward layer in Transformer model')`<br/>
    `group.add_argument('-head_count', type=int, default=8,
                       help='Number of heads for Transformer MultiHeadAttention')`<br/>
<br/>

2.  Explicitly pass these flags to the Encoder and Decoder initialization in onmt/ModelConstructor.py make_encoder and make_decoder functions for transformer case:<br/>
<br/>

       `return TransformerEncoder(opt.enc_layers, opt.rnn_size,
                                  opt.dropout, embeddings,
                                  opt.ff_size, opt.head_count) #NEW pass flag values for ff_size, head_count to Encoder init` <br/>
<br/>

        `return TransformerDecoder(opt.dec_layers, opt.rnn_size,
                                  opt.global_attention, opt.copy_attn,
                                  opt.dropout, embeddings,
                                  opt.ff_size, opt.head_count) #NEW pass flag values for ff_size, head_count to Decoder init`<br/>
<br/>

3.  Pass these flags along inside TransformerEncoder init to TransformerEncoderLayer init and TransformerDecoder init to TransformerDecoderLayer init, where the TransformerEncoderLayer and TransformerDecoderLayer init functions are already parameterized to receive them, in onmt/modules/Transformer.py:<br/>
<br/>

Encoder:<br/>
<br/>

    `def __init__(self, num_layers, hidden_size,
                 dropout, embeddings,
                 ff_size, head_count): #NEW pass ff_size and head_count to Encoder init`<br/>
<br/>

        `self.transformer = nn.ModuleList(
            [TransformerEncoderLayer(hidden_size, dropout, hidden_size=ff_size, head_count=head_count) for _ in range(num_layers)]) #NEW pass ff_size and head_count to layer init`<br/>
<br/>

Decoder:<br/>
<br/>

   `def __init__(self, num_layers, hidden_size, attn_type,
                 copy_attn, dropout, embeddings,
                 ff_size, head_count): #NEW pass ff_size and head_count to decoder init`<br/>
        <br/>

        `self.transformer_layers = nn.ModuleList(
            [TransformerDecoderLayer(hidden_size, dropout, hidden_size=ff_size, head_count=head_count) for _ in range(num_layers)]) #NEW pass ff_size and head_count to layer init `<br/>

**Comment by vince62s** [comment]
done in #798

**Comment by David-Levinthal** [comment]
cool
:-)
